js.go

```diff
@@ -1050,11 +1066,52 @@ func (js *js) PublishMsgAsync(m *Msg, opts ...PubOpt) (PubAckFuture, error) {
 			case <-js.asyncStall():
 			case <-time.After(stallWait):
 				js.clearPAF(id)
-				return nil, errors.New("nats: stalled with too many outstanding async published messages")
+				return nil, ErrTooManyStalledMsgs
 			}
 		}
+		if js.opts.ackTimeout > 0 {
+			paf.timeout = time.AfterFunc(js.opts.ackTimeout, func() {
+				js.mu.Lock()
+				defer js.mu.Unlock()
+
+				if _, ok := js.pafs[id]; !ok {
+					// paf has already been resolved
+					// while waiting for the lock
+					return
+				}
+
+				// ack timed out, remove from pending acks
+				delete(js.pafs, id)
```

Comment:
another thing that we could consider is adding `Cancel` to the PubAcks interface with a closure over `(js.pafs, id)` so that user can control releasing these resources when needed

Reply:
this way the pub acks would behave more like futures, return the PubAck and in case the client no longer interested in the response they can then cancel to release the resources, similar to the context package

Reply:
Then the other part that could be missing in the future is being able to check whether the future is `IsDone()` or not plus its internal timestamp check if its stale, we could then be collecting all the futures inflight for example. I think surfacing that may give folks to manage this on their own in case they want to be more efficiently